homeassistant/components/hue/__init__.py

```diff
@@ -105,16 +104,46 @@ async def async_setup_entry(
     host = entry.data["host"]
     config = hass.data[DATA_CONFIGS].get(host)
 
-    if config is None:
-        allow_unreachable = entry.data.get(
-            CONF_ALLOW_UNREACHABLE, DEFAULT_ALLOW_UNREACHABLE
-        )
-        allow_groups = entry.data.get(CONF_ALLOW_HUE_GROUPS, DEFAULT_ALLOW_HUE_GROUPS)
-    else:
-        allow_unreachable = config[CONF_ALLOW_UNREACHABLE]
-        allow_groups = config[CONF_ALLOW_HUE_GROUPS]
+    # Migrate allow_unreachable from config entry data to config entry options
+    if (
+        CONF_ALLOW_UNREACHABLE not in entry.options
+        and CONF_ALLOW_UNREACHABLE in entry.data
+        and entry.data[CONF_ALLOW_UNREACHABLE] != DEFAULT_ALLOW_UNREACHABLE
+    ):
+        options = {
+            **entry.options,
+            CONF_ALLOW_UNREACHABLE: entry.data[CONF_ALLOW_UNREACHABLE],
+        }
+        data = entry.data.copy().pop(CONF_ALLOW_UNREACHABLE)
+        hass.config_entries.async_update_entry(entry, data=data, options=options)
+
+    # Migrate allow_hue_groups from config entry data to config entry options
+    if (
+        CONF_ALLOW_HUE_GROUPS not in entry.options
+        and CONF_ALLOW_HUE_GROUPS in entry.data
+        and entry.data[CONF_ALLOW_HUE_GROUPS] != DEFAULT_ALLOW_HUE_GROUPS
+    ):
+        options = {
+            **entry.options,
+            CONF_ALLOW_HUE_GROUPS: entry.data[CONF_ALLOW_HUE_GROUPS],
+        }
+        data = entry.data.copy().pop(CONF_ALLOW_HUE_GROUPS)
+        hass.config_entries.async_update_entry(entry, data=data, options=options)
+
+    # Overwrite from YAML configuration
+    if config is not None:
+        options = {}
+        if CONF_ALLOW_HUE_GROUPS in config:
+            options[CONF_ALLOW_HUE_GROUPS] = config[CONF_ALLOW_HUE_GROUPS]
```

Comment:
Should we make these checks conditional?

```suggestion
        if CONF_ALLOW_HUE_GROUPS in config and config[CONF_ALLOW_HUE_GROUPS] != entry.options[CONF_ALLOW_HUE_GROUPS]:
            options[CONF_ALLOW_HUE_GROUPS] = config[CONF_ALLOW_HUE_GROUPS]
```